src/CountIterableCapableTrait.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+namespace Dhii\Iterator;
+
+use Traversable;
+use IteratorAggregate;
+use Countable;
+
+/**
+ * Functionality for counting elements in an iterable.
+ *
+ * @since [*next-version*]
+ */
+trait CountIterableCapableTrait
+{
+    /**
+     * Counts the elements in an iterable.
+     *
+     * Is optimized to retrieve count from values that support it.
+     * - If array, will count in regular way using count();
+     * - If {@see Countable}, will do the same;
+     * - If {@see IteratorAggregate}, will drill down into internal iterators
+     * until the first {@see Countable} is encountered, in which case the same
+     * as above will be done.
+     * - In any other case, will apply {@see iterator_count()}, which means
```

Comment:
Ideally also mention that in such a scenario, the iterator pointer of the iterator is not guaranteed to be preserved.